message/messaging_service.cc

```diff
@@ -866,6 +869,8 @@ shared_ptr<messaging_service::rpc_protocol_client_wrapper> messaging_service::ge
     opts.reuseaddr = true;
     opts.isolation_cookie = _scheduling_info_for_connection_index[idx].isolation_cookie;
 
+    assert(!must_encrypt || _credentials);
```

Comment:
Is this assert() the right thing to do? Could some configuration error (e.g., missing certificate) cause an assertion failure here? It's very frustrating to get an assertion failure instead of a configuration-error message :-(

In general, I'm not a fan of assert(), so let's think carefully if we really need one here. If it's really a bug (and not a simple configuration error), why not an on_internal_error() instead?

Reply:
Again, using an assert as asserts are intended (extra sanity checks, not runtime conditions -> expected/valid exceptions), yes, this is correct. 
Broken TLS config will cause initialization `init` to fail -> startup should fail. If we get here and must_encrypt is true but no certs are loaded, we've again torn the carpet out of class init assumptions, and assert is the way to handle such. 
Again: if the condition here hits, it is not an "internal error" (unforseen, but possible error), but broken application logic again -> fundamental code issue.

Reply:
That's not what exactly what I see, but maybe I'm missing something? It seems that init() under some conditions (_!credential_builders, I'm not sure what this means) leaves _credentials unset. I don't see in init() any code that checks if _credentials is unset prints a nice message and throws, or anything like that...
In any case, I don't know where you got the on_internal_error() means "unforseen but possible error". It strictly means a logic error in the code, we reached an impossible situation and should report a bug. HOWEVER, it doesn't mean we need to crash Scylla. We can throw an exception, which in some cases breaks just the current operation, and in the startup case will abort Scylla's initialization, but will do so in a relatively clean way - not an ugly core dump.

Reply:
That is my point. This is a coredump and rethink situation. Not a "throw an exception, which in some cases breaks just the current operation" one. This is a fundamental error in the code structure, not "oops, something failed to initialize".
If init does not set credentials, then must_encrypt should never be true. Because we must have either finished init such that credentials are created, or we've already reported the error, failed main init and exited.

Reply:
> If init does not set credentials, then must_encrypt should never be true.

Is there an non-asserting check for that e.g. in main or messaging_service constructor? Adding a reference to it near assertion would be very helpful